netty/src/test/java/io/grpc/netty/ProtocolNegotiatorsTest.java

```diff
@@ -300,6 +324,128 @@ public ChannelCredentials withoutBearerTokens() {
     assertThat(result.negotiator).isNull();
   }
 
+  private InternalChannelz.Tls expectSuccessfulHandshake(
+      ChannelCredentials channelCreds, ServerCredentials serverCreds) throws Exception {
+    MockServerListener serverListener = new MockServerListener();
+    ClientTransportFactory clientFactory = NettyChannelBuilder
+        // Although specified here, address is ignored because we never call build.
+        .forAddress("localhost", 0, channelCreds)
+        .buildTransportFactory();
+    InternalServer server = NettyServerBuilder
+        .forPort(0, serverCreds)
+        .buildTransportServers(Collections.<ServerStreamTracer.Factory>emptyList());
+    server.start(serverListener);
+
+    ManagedClientTransport.Listener clientTransportListener =
+        mock(ManagedClientTransport.Listener.class);
+    ManagedClientTransport client = clientFactory.newClientTransport(
+        server.getListenSocketAddress(),
+        new ClientTransportFactory.ClientTransportOptions()
+          .setAuthority(TestUtils.TEST_SERVER_HOST),
+        mock(ChannelLogger.class));
+    callMeMaybe(client.start(clientTransportListener));
+    verify(clientTransportListener, timeout(TIMEOUT_SECONDS * 1000)).transportReady();
+    InternalChannelz.SocketStats stats = serverListener.transports.poll().getStats().get();
+    assertThat(stats.security).isNotNull();
+    assertThat(stats.security.tls).isNotNull();
+
+    client.shutdownNow(Status.UNAVAILABLE.withDescription("trash it"));
+    server.shutdown();
+    assertTrue(
+        serverListener.waitForShutdown(TIMEOUT_SECONDS * 1000, TimeUnit.MILLISECONDS));
+    verify(clientTransportListener, timeout(TIMEOUT_SECONDS * 1000)).transportTerminated();
+    clientFactory.close();
+    return stats.security.tls;
+  }
+
+  @Test
+  public void from_tls_basic() throws Exception {
+    ServerCredentials serverCreds = TlsServerCredentials.create(server1Cert, server1Key);
+    ChannelCredentials channelCreds = TlsChannelCredentials.newBuilder()
+        .trustManager(caCert)
+        .build();
+    InternalChannelz.Tls tls = expectSuccessfulHandshake(channelCreds, serverCreds);
+    assertThat(tls.remoteCert).isNull();
+  }
+
+  @Test
+  public void from_tls_mtls_require() throws Exception {
+    ServerCredentials serverCreds = TlsServerCredentials.newBuilder()
+        .keyManager(server1Cert, server1Key)
+        .trustManager(caCert)
+        .clientAuth(TlsServerCredentials.ClientAuth.REQUIRE)
```

Comment:
How about a negative test with `ClientAuth.REQUIRE` where the client doesn't present a certificate?